### PR TITLE
Cast CUPY_SEED environment variable to uint64

### LINKED ISSUE #766: Setting CUPY_SEED or CHAINER_SEED raises a TypeError
If `CUPY_SEED` or `CHAINER_SEED` is set, `cupy.random` raises a TypeError. I guess it is due to this commit: https://github.com/cupy/cupy/commit/03b2145fbec95166832de56506d8fbc256406e39#diff-bdf8bf88a45064e9d8d316ebd0d0f6fb

```
$ python -c "import cupy; print(cupy.__version__)"
2.1.0
$ python -c "import cupy; cupy.random.rand()"
$ CUPY_SEED=0 python -c "import cupy; cupy.random.rand()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/sample.py", line 33, in rand
    return random_sample(size=size, dtype=dtype)
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/sample.py", line 153, in random_sample
    rs = generator.get_random_state()
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/generator.py", line 412, in get_random_state
    rs = RandomState(seed)
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/generator.py", line 47, in __init__
    self.seed(seed)
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/generator.py", line 243, in seed
    seed = numpy.asarray(seed).astype(numpy.uint64, casting='safe')
TypeError: Cannot cast array from dtype('<U1') to dtype('uint64') according to the rule 'safe'
$ CHAINER_SEED=0 python -c "import cupy; cupy.random.rand()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/sample.py", line 33, in rand
    return random_sample(size=size, dtype=dtype)
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/sample.py", line 153, in random_sample
    rs = generator.get_random_state()
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/generator.py", line 412, in get_random_state
    rs = RandomState(seed)
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/generator.py", line 47, in __init__
    self.seed(seed)
  File "/home/fujita/.pyenv/versions/mbrl/lib/python3.6/site-packages/cupy/random/generator.py", line 243, in seed
    seed = numpy.asarray(seed).astype(numpy.uint64, casting='safe')
TypeError: Cannot cast array from dtype('<U1') to dtype('uint64') according to the rule 'safe'
```

### PR DESCRIPTION
**Pull request by toslunar**
Fixes #766

### CODE REVIEW

File: tests/cupy_tests/random_tests/test_generator.py [hunk 1]
```diff
@@ -706,21 +706,21 @@ def test_get_random_state_no_cupy_no_chainer_seed(self):
 
     def test_get_random_state_no_cupy_with_chainer_seed(self):
         os.environ.pop('CUPY_SEED', None)
-        os.environ['CHAINER_SEED'] = '5'
+        os.environ['CHAINER_SEED'] = str(5)
```
Comment (niboshi) [inline tests/cupy_tests/random_tests/test_generator.py]: This use of `str` is not necessary. Could you revert it? (same for other changes)

**Comment by niboshi** [comment]
Thank you for the fix.
In this test, `mock` seems to have hidden the bug.

We should avoid using `mock` in this test.
Instead, we should write a test to verify the seed by sampling some random distribution twice, one with environment variables, the other with explicit seed, and compare the sampled numbers.

Could you write this test?
Or if you want, I can merge in the current status and make another PR by myself.

**Comment by toslunar** [comment]
I added tests that call `get_random_state` twice with the same seed or with different seeds.  I used `reset_states` to delete cached `RandomState`.

**Comment by niboshi** [comment]
Thank you for writing tests.
However, what we should check is consistency between a seed set by environment variables vs explicitly given seed (reproducibility should be checked in another place). So the simplified logic would be:

```
os.environ['CUPY_SEED'] = '7'
os.environ['CHAINER_SEED'] = '8'
x1 = cupy.random.generate_some_random()

cupy.random.seed(7)
x2 = cupy.random.generate_some_random()

assert x1 == x2
```

We need to test against possible combination of `CHAINER_SEED` and `CUPY_SEED`.
Also, `os.environ` must be rolled back after test, so that it will not affect other tests.

I think it's better to rewrite tests in `TestGetRandomState2`, because current tests using mock are not very meaningful anyway.

**Comment by toslunar** [comment]
- I understood the need of the rollback of `os.environ`, and fixed it.
- A mock test remains for the case neither CUPY_SEED nor CHAINER_SEED is given.
- I kept reproducibility test since I'd like to be sure the testing method (i.e. comparing output randoms) is valid.

**Comment by niboshi** [comment]
* I don't still think mock is necessary. How about sampling twice without environment variables and seeds, and asserting they're different numbers? This way you can confirm they're sampled from different seeds.
* Reproducibility of random numbers are checked here: https://github.com/toslunar/cupy/blob/5bc4d6eac11eb64d2beef7f0601ad995878f407c/tests/cupy_tests/random_tests/test_generator.py#L31

**Comment by toslunar** [comment]
Thanks. Fixed.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM, Thank you!!

**Pull request merged**